server/gitrest/packages/gitrest-base/src/utils/repositoryManagerFactoryBase.ts

```diff
@@ -0,0 +1,234 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { E_TIMEOUT, Mutex, MutexInterface, withTimeout } from "async-mutex";
+import { NetworkError } from "@fluidframework/server-services-client";
+import { Lumberjack } from "@fluidframework/server-services-telemetry";
+import { IExternalStorageManager } from "../externalStorageManager";
+import * as helpers from "./helpers";
+import {
+    IRepositoryManagerFactory,
+    IRepositoryManager,
+    IFileSystemManager,
+    IFileSystemManagerFactory,
+    IRepoManagerParams,
+    IStorageDirectoryConfig,
+    BaseGitRestTelemetryProperties,
+    Constants,
+} from "./definitions";
+
+type RepoOperationType = "create" | "open";
+
+export abstract class RepositoryManagerFactoryBase<TRepo> implements IRepositoryManagerFactory {
+    // Cache repositories to allow for reuse
+    private readonly repositoryCache = new Map<string, TRepo>();
```

Comment:
Request that has nothing to do with your changes: while you're at it, could you make this `protected`? In Spork, I have an extension of RepoManagerFactory that deletes repos (they are temporary for computation) and I need to be able to clear the cache to avoid thinking a repo exists when it does not. Without this being `protected` I have to duplicate a lot of the code